File: android/src/main/java/com/yandexmobileads/util/factory/AdRequestConfigurationFactory.kt
package com.yandexmobileads.util.factory

import com.facebook.react.bridge.ReadableMap
import com.yandex.mobile.ads.common.AdRequestConfiguration
import com.yandex.mobile.ads.common.Gender
import com.yandex.mobile.ads.common.AdTheme
import com.yandexmobileads.common.Constants
import com.yandexmobileads.util.Converter

internal class AdRequestConfigurationFactory(
    private val locationFactory: LocationFactory = LocationFactory(),
    private val converter: Converter = Converter(),
) {
    fun createFromReadableMap(readableMap: ReadableMap): AdRequestConfiguration? {
        val adUnitId = readableMap.getString(Constants.Key.AD_UNIT_ID) ?: return null

        val adRequestConfiguration = AdRequestConfiguration.Builder(adUnitId)

        adRequestConfiguration.setAge(readableMap.getString(Constants.Key.AGE))
        adRequestConfiguration.setLocation(
            locationFactory.createFromReadableMap(
                readableMap.getMap(Constants.Key.LOCATION)
            )
        )
        adRequestConfiguration.setContextQuery(readableMap.getString(Constants.Key.CONTEXT_QUERY))
        adRequestConfiguration.setContextTags(
            converter.convertReadableArrayToList(
                readableMap.getArray(
                    Constants.Key.CONTEXT_TAGS
                )
            )
        )
        adRequestConfiguration.setGender(
            when (readableMap.getString(Constants.Key.GENDER)) {
                Constants.Gender.MALE -> Gender.MALE
                Constants.Gender.FEMALE -> Gender.FEMALE
                else -> null
            }
        )
        adRequestConfiguration.setPreferredTheme(
            when (readableMap.getString(Constants.Key.AD_THEME)) {
                Constants.AdTheme.LIGHT -> AdTheme.LIGHT
                Constants.AdTheme.DARK -> AdTheme.DARK
                else -> null
            }
        )
        adRequestConfiguration.setBiddingData(readableMap.getString(Constants.Key.BIDDING_DATA))
        adRequestConfiguration.setParameters(
            converter.convertReadableMapToMap(
                readableMap.getMap(Constants.Key.PARAMETERS)
            )
        )

        return adRequestConfiguration.build()
    }
}

File: ios/Utils/Factory/AdRequestConfigurationFactory.swift
import Foundation
import YandexMobileAds

class AdRequestConfigurationFactory {
    private let locationFactory: LocationFactory
    
    convenience init() {
        let locationFactory = LocationFactory()
        self.init(locationFactory: locationFactory)
    }
    
    init(locationFactory: LocationFactory) {
        self.locationFactory = locationFactory
    }
    
    func createFromDictionary(_ dict: [String: Any]) -> AdRequestConfiguration? {
        guard let adUnitId = dict[Constants.Key.adUnitId] as? String else {
            return nil
        }
        
        let adRequestConfiguration = AdRequestConfiguration(adUnitID: adUnitId)
        let mutableConfiguration = adRequestConfiguration.mutableConfiguration
        
        mutableConfiguration.location = locationFactory.createFromDictionary(dict[Constants.Key.location] as? [String: Any])
        mutableConfiguration.contextQuery = dict[Constants.Key.contextQuery] as? String
        mutableConfiguration.contextTags = dict[Constants.Key.contextTags] as? [String]
        mutableConfiguration.biddingData = dict[Constants.Key.biddingData] as? String
        mutableConfiguration.parameters = dict[Constants.Key.parameters] as? [String: String]
        
        if let ageString = dict[Constants.Key.age] as? String,
           let ageInt = Int(ageString) {
            mutableConfiguration.age = NSNumber(value: ageInt)
        }
        
        if let gender = dict[Constants.Key.gender] as? String {
            switch gender {
            case Constants.Gender.male:
                mutableConfiguration.gender = kYMAGenderMale
            case Constants.Gender.female:
                mutableConfiguration.gender = kYMAGenderFemale
            default:
                mutableConfiguration.gender = nil
            }
        }
        
        mutableConfiguration.adTheme = .unspecified
        if let adTheme = dict[Constants.Key.adTheme] as? String {
            switch adTheme {
            case Constants.AdTheme.light:
                mutableConfiguration.adTheme = .light
            case Constants.AdTheme.dark:
                mutableConfiguration.adTheme = .dark
            default:
                break
            }
        }
        
        return mutableConfiguration
    }
}
